docs/extensibility.md

```diff
@@ -27,19 +30,118 @@ To create an Upgrade Assistant extension, you will need to start with a manifest
 
 The ExtensionServiceProviders element of the extension manifest contains an array of assemblies that the Upgrade Assistant should look in for implementations of `Microsoft.DotNet.UpgradeAssistant.Extensions.IExtensionServiceProvider`. At runtime, Upgrade Assistant will load any assemblies listed in the ExtensionServiceProviders array (paths are relative to the extension manifest's location) and instantiate an public implementations of `IExtensionServiceProvider` found in those assemblies. The `IExtensionServiceProvider` instances will then be used to register services in Upgrade Assistant's dependency injection container. Common services that an extension might register include:
 
-1. Custom upgrade steps (inheriting from `UpgradeStep`). Any upgrade steps registered by an `IExtensionServiceProvider` in an assembly listed in the ExtensionServiceProviders array will be included in the upgrade steps the Upgrade Assistant executes. In this way, extenders can add their own custom steps to the upgrade pipeline.
-2. Roslyn analyzers and code fix providers. Upgrade Assistant's source updater step looks in the dependency injection container for any analyzers with associated code fix providers and will include them in the sub-steps to the source updater step. So, by registering their own Roslyn analyzers and code fix providers, extenders can customize the source update steps used by Upgrade Assistant.
-3. `IConfigUpdater` implementations. Upgrade Assistant's config updater step uses any registered `IConfigUpdater` services to make project updates based on config files (app.config, web.config). Therefore, by registering their own `IConfigUpdater` implementations, extenders can customize how config-related upgrades are made by Upgrade Assistant.
-4. `IPackageReferencesAnalyzer` implementations. For more information on how the package updater step works, see the next section of this document. If providing custom package mapping configuration is insufficient, however, extenders can register their own implementations of `IPackageReferenceAnalzyer` to more completely customize how NuGet package references are updated.
-5. Any other services that might be needed by Upgrade Assistant steps (either the default steps or those added by extensions). Extenders can register services that their own upgrade steps will need or services that will be used by other upgrade steps and Upgrade Assistant will make sure any services registered in an `IExtensionServiceProvider` implementation will be made available at runtime.
+## Steps
+Custom upgrade steps (inheriting from `UpgradeStep`). Any upgrade steps registered by an `IExtensionServiceProvider` in an assembly listed in the ExtensionServiceProviders array will be included in the upgrade steps the Upgrade Assistant executes. In this way, extenders can add their own custom steps to the upgrade pipeline.
```

Comment:
personal preference, raised to see if it speaks to anyone else. feel free to ignore

I'd want to read that `the upgrade-assistant will scan for extensions that exist in class libraries`. Those words explicitly tell me what to expect about the extension model, similar to MEF behavior.

Reply:
I see what you're getting at @KSchlobohm, but the scanning that happens is when UA scans libraries listed as ExtensionServiceProviders for `IExtensionServiceProvider` implementations and uses those to register services with DI.

So, a more correct rephrasing of your statement might be:

> Customer upgrade steps (inheriting from `UpgradeStep`) can be used by registering them in the dependency injection container. Upgrade Assistant will scan class libraries listed in extensions' ExtensionServiceProviders for implementations of `IExtensionServiceProvider` which can register the UpgradeSteps (or other services), as needed.

That ends up being more about the general extension mechanics than Upgrade Steps in particular, though, so maybe it belongs earlier in the docs.

Reply:
I've moved the section about IExtensionServiceProvider to be near the top before any specific thing you could add